### PR TITLE
[MRG] ENH Add example comparing the distribution of all scaling preprocessor

### PR DESCRIPTION
**Pull request by raghavrv**
Main PR: https://github.com/scikit-learn/scikit-learn/pull/8363

I have some initial plots comparing different scalers in the california housing dataset...
I binarized the targets to help visualize it better...

For quantile normalizer

image

image

@glemaitre @tguillemot @ogrisel @dengemann @jnothman

### CODE REVIEW

**Comment by glemaitre** [comment]
@raghavrv can you plot the robust scaler?

**Comment by dengemann** [comment]
That looks pretty cool! Indeed it seems to support the point made by @ogrisel about the potentially decorrelating nature of this non-linear transform.

**Comment by ogrisel** [comment]
I don't understand why the normalized data looks discretized on the y axis. Aren't we linearly interpolating points between quantiles?

**EDIT** by @raghavrv (This was for an old deleted plot)

**Comment by glemaitre** [comment]
@ogrisel Checking with @raghavrv IRL, it was a mistake.

**Comment by raghavrv** [comment]
@ogrisel Sorry I plotted the wrong features... ~~Here~~ Have updated at the PR description the plots for all the scalers / normalizers. You can see our QN performing great!

**Comment by raghavrv** [comment]
Note that I didn't binarize it this time and instead used matplotlib's colormap...

**Comment by ogrisel** [comment]
The plots looks great!

I find it weird that robust scaler is yielding such a flat profile. How many "outliers" are there on the y axis? I would have assumed that IQR scaling should have performed better.

**Comment by ogrisel** [comment]
Please add x and y labels on the first 2 plots (data without transform and after zoom on non-outliers).

**Comment by raghavrv** [comment]
> How many "outliers" are there on the y axis? I would have assumed that IQR scaling should have performed better.

It scales the outliers by the IQR, so the outliers will still be outliers in the scaled data and hence the flat profile... We could zoom in like it is done [here](http://scikit-learn.org/stable/auto_examples/preprocessing/plot_robust_scaling.html#sphx-glr-auto-examples-preprocessing-plot-robust-scaling-py), but it would be unfair to the other scalers I think....

**Comment by jnothman** [comment]
Maybe best to just show 0-1 range after scaling.

**Comment by jnothman** [comment]
or show 0-1 and mark outliers along edges with x's

**Comment by ogrisel** [comment]
> or show 0-1 and mark outliers along edges with x's

I am not sure how showing the outliers would work, but I like the idea of zooming on the [0.01-0.99] quantiles for all the transforms (in addition to the original, unzoomed plot for each transform).

**Comment by raghavrv** [comment]
@ogrisel @jnothman Okay this is the re-updated plot :)

image

**Comment by ogrisel** [comment]
Very nice!

**Comment by ogrisel** [comment]
Maybe you could remove the color bar from the first plots and only display it for the last plot as they all use the same color scale.

**Comment by dengemann** [comment]
Great work! It looks very helpful and clear.

**Comment by tguillemot** [comment]
Nice work @raghavrv !

File: examples/preprocessing/plot_all_scaling.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                      X_label="Median Income", y_label="Number of households")
+
+    X0_min, X0_99th_pc = np.percentile(X[:, 0], [0, 99])
+    X1_min, X1_99th_pc = np.percentile(X[:, 1], [0, 99])
+
+    non_outliers = np.all(X < [X0_99th_pc, X1_99th_pc], axis=1)
+    axes = subplots[offset + 2: offset + 4] + subplots[offset + 6: offset + 8]
+    plot_distribution(axes, X[non_outliers], y[non_outliers], hist_nbins=50,
+                      plot_title=(title +
+                                  "\n(Zoomed-in at quantile range [0, 99))"),
```
Comment (jnothman) [inline examples/preprocessing/plot_all_scaling.py]: quantile -> percentile.

Also, I think parentheses enclosing these subtitles can be removed.

**Comment by jnothman** [comment]
Great stuff!

File: examples/preprocessing/plot_all_scaling.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/python
```
Comment (glemaitre) [inline examples/preprocessing/plot_all_scaling.py]: `#!/usr/bin/env python` should be better

File: examples/preprocessing/plot_all_scaling.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+"""
+=============================================================
+Compare the effect of different scalers on data with outliers
+=============================================================
+
+The feature 0 and feature 5 of california housing dataset contains large
+outliers that can make visualization of the data difficult.
+
+Also linear models like :class:`sklearn.linear_model.SVM` require data which is
+normalized to the range [-1, 1].
```
Comment (ogrisel) [inline examples/preprocessing/plot_all_scaling.py]: approximately normalized to the [-1, 1] or [0, 1] range, or at the very least have all the features on the same scale.

File: examples/preprocessing/plot_all_scaling.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+from sklearn.preprocessing import Normalizer
+from sklearn.preprocessing.data import QuantileNormalizer
+
+from sklearn.datasets import fetch_california_housing
+
+dataset = fetch_california_housing()
+X_full, y_full = dataset.data, dataset.target
+
+# Take only 2 features to make visualization easier
+# Feature 0 has a tapering distribution of outliers
```
Comment (ogrisel) [inline examples/preprocessing/plot_all_scaling.py]: Feature of 0 has a long tail distribution. Feature 5 has a few but very large outliers.

File: examples/preprocessing/plot_all_scaling.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    heatmap_legend_ax = plt.subplot(gs[offset+8: offset+12])
+    norm = mpl.colors.Normalize(y_full.min(), y_full.max())
+    mpl.colorbar.ColorbarBase(heatmap_legend_ax, cmap=cm.plasma_r,
+                              norm=norm, orientation='horizontal',
+                              label='color mapping for values of y')
+
+    # Blank space to avoid overlapping of plots;
+    # plt.tight_layout does not work with gridspec, height of this row is
+    # adjusted at `height_ratios` param given to `GridSpec`
+    plt.subplot(gs[offset+12: offset+16]).axis('off')
```
Comment (ogrisel) [inline examples/preprocessing/plot_all_scaling.py]: cosmetics: `gs[offset + 12:offset + 16]` or `gs[offset+12:offset+16]` but no spacing around the slicing operator `:` and especially not asymmetric, otherwise it looks like the literal notation of a dict.

**Comment by raghavrv** [comment]
Thanks all for the comments :)

Any idea on how the doc can be improved? Should it have a note for each scaler/norm. stating it's use case?

(Have updated the above comment with the new plot)

**Comment by ogrisel** [comment]
> Any idea on how the doc can be improved? Should it have a note for each scaler/norm. stating it's use case?

Yes I think that would be useful to give a high level analysis of the behavior on each w.r.t. this specific 2D dataset.

**Comment by jnothman** [comment]
I'm now wondering if there's any value showing the transformations on held out data

**Comment by ogrisel** [comment]
> I'm now wondering if there's any value showing the transformations on held out data

They should look very similar to the data on the training set, right? Adding plots for held out data would render this example quite complex. I am not sure it's worth the additional cognitive load.

**Comment by jnothman** [comment]
Happy to leave it. Yes it should look similar except for minmax.

**Comment by glemaitre** [comment]
Merged!!!

**Comment by raghavrv** [comment]
Thanks for merging! :tada: I'll edit the docs in the main branch...
